---
title: "amplicon_analysis"
output: html_document
---
# Package Installation

```
install.packages("BiocManager")
BiocManager::install("dada2", version = "3.12")
```

```
source("https://bioconductor.org/biocLite.R")
biocLite("ShortRead")
install.packages("dplyr")
install.packages("tidyr")
install.packages("Hmisc")
install.packages("ggplot2")
install.packages("plotly")
```
If you are running it on your own computer (runs slower!):

Download this tutorial from github. Go to the homepage, and click the green "Clone or download" button. Then click "Download ZIP", to save it to your computer. Unzip the file to access the R-script.

idemp can be found here: https://github.com/yhwu/idemp

cutadapt can be installed from here: https://cutadapt.readthedocs.io/en/stable/installation.html

Download the dada2-formatted reference database of your choice. 
Link to download here: https://benjjneb.github.io/dada2/training.html

# Setup
## Load DADA2 and required packages

```{r}
library(dada2); packageVersion("dada2") # the dada2 pipeline
## [1] '1.10.1'
library(ShortRead); packageVersion("ShortRead") # dada2 depends on this
## [1] '1.38.0'
library(dplyr); packageVersion("dplyr") # for manipulating data
## [1] '0.8.0.1'
library(tidyr); packageVersion("tidyr") # for creating the final graph at the end of the pipeline
## [1] '0.8.2'
library(Hmisc); packageVersion("Hmisc") # for creating the final graph at the end of the pipeline
## [1] '4.2.0'
library(ggplot2); packageVersion("ggplot2") # for creating the final graph at the end of the pipeline
## [1] '3.1.0'
library(plotly); packageVersion("plotly") # enables creation of interactive graphs, especially helpful for quality plots
## [1] '4.8.0'

```

## Set up pathway to idemp (demultiplexing tool) and test
```{r}
idemp <- "/Users/Tristan/active_repos/soil_amplicon_analysis_21/idemp/idemp" # CHANGE ME if not on microbe
system2(idemp) # Check that idemp is in your path and you can run shell commands from R
```

## Set up pathway to cutadapt (primer trimming tool) and test
```{r}
cutadapt <- "/Users/Tristan/anaconda3/envs/cutadaptenv/lib/python3.9/site-packages/cutadapt" # CHANGE ME if not on microbe
system2(cutadapt, args = "--version") # Check by running shell command from R
```

## Set path to shared data folder and contents [change me!]
```{r}
data.fp <- "2021_06_21_DebOREI_Sierra_16s/210621_Henley_Fierer-1268"
# List all files in shared folder to check path
list.files(data.fp)
```

## Set file paths for barcodes file, map file, and fastq files [change me!]
Barcodes need to have 'N' on the end of each 12bp sequence for compatibility!

```{r}
barcode.fp <- file.path(data.fp, "Deb_OREI_Plates9-12_16s_Barcodes.txt") # .txt file: barcode </t> sampleID
map.fp <- file.path(data.fp, "Deb_OREI_Plates9-12_16s_MappingFile.txt")
I1.fp <- file.path(data.fp, "Undetermined_S0_L001_I1_001.fastq.gz") 
R1.fp <- file.path(data.fp, "Undetermined_S0_L001_R1_001.fastq.gz") 
R2.fp <- file.path(data.fp, "Undetermined_S0_L001_R2_001.fastq.gz") 
```

>NOTE: idemp relies on having a match in length between the index file and and the barcode sequences. Since the index file usually includes a extra linker basepair (making it 13bp long), you should append the barcode sequences with "N" to make sure each is 13bp long. If you are not sure of the length of index reads, check with the sequencing center. If your index reads are 12bp long, you do NOT need to add an "N".

## Set up file paths in YOUR directory where you want data [change me!]
You do not need to create the subdirectories but they are nice to have for organizational purposes.

```{r}
project.fp <- "/Users/Tristan/active_repos/soil_amplicon_analysis_21" # CHANGE ME to project directory; don't append with a "/"

# Set up names of sub directories to stay organized
preprocess.fp <- file.path(project.fp, "01_preprocess")
    demultiplex.fp <- file.path(preprocess.fp, "demultiplexed")
    filtN.fp <- file.path(preprocess.fp, "filtN")
    trimmed.fp <- file.path(preprocess.fp, "trimmed")
filter.fp <- file.path(project.fp, "02_filter") 
table.fp <- file.path(project.fp, "03_tabletax") 
```

# Pre-processing data for dada2 - demultiplex, remove sequences with Ns, cutadapt
## Call the demultiplexing script
Demultiplexing splits your reads out into separate files based on the barcodes associated with each sample.

```{r}
flags <- paste("-b", barcode.fp, "-I1", I1.fp, "-R1", R1.fp, "-R2", R2.fp, "-o", demultiplex.fp) 
system2(idemp, args = flags) 
```


## Clean up the output from idemp
```{r}
# Change names of unassignable reads so they are not included in downstream processing
unassigned_1 <- paste0("mv", " ", demultiplex.fp, "/Undetermined_S0_L001_R1_001.fastq.gz_unsigned.fastq.gz",
                       " ", demultiplex.fp, "/Unassigned_reads1.fastq.gz")
unassigned_2 <- paste0("mv", " ", demultiplex.fp, "/Undetermined_S0_L001_R2_001.fastq.gz_unsigned.fastq.gz", 
                       " ", demultiplex.fp, "/Unassigned_reads2.fastq.gz")
system(unassigned_1)
system(unassigned_2)

# Rename files - use gsub to get names in order!
R1_names <- gsub(paste0(demultiplex.fp, "/Undetermined_S0_L001_R1_001.fastq.gz_"), "", 
                 list.files(demultiplex.fp, pattern="R1", full.names = TRUE))
file.rename(list.files(demultiplex.fp, pattern="R1", full.names = TRUE), 
            paste0(demultiplex.fp, "/R1_", R1_names))

R2_names <- gsub(paste0(demultiplex.fp, "/Undetermined_S0_L001_R2_001.fastq.gz_"), "", 
                 list.files(demultiplex.fp, pattern="R2", full.names = TRUE))
file.rename(list.files(demultiplex.fp, pattern="R2", full.names = TRUE),
            paste0(demultiplex.fp, "/R2_", R2_names))

# Get full paths for all files and save them for downstream analyses
# Forward and reverse fastq filenames have format: 
fnFs <- sort(list.files(demultiplex.fp, pattern="R1_", full.names = TRUE))
fnRs <- sort(list.files(demultiplex.fp, pattern="R2_", full.names = TRUE))
```

## Pre-filter to remove sequence reads with Ns
Ambiguous bases will make it hard for cutadapt to find short primer sequences in the reads. To solve this problem, we will remove sequences with ambiguous bases (Ns)

```{r}
# Name the N-filtered files to put them in filtN/ subdirectory
fnFs.filtN <- file.path(preprocess.fp, "filtN", basename(fnFs))
fnRs.filtN <- file.path(preprocess.fp, "filtN", basename(fnRs))

# Filter Ns from reads and put them into the filtN directory
filterAndTrim(fnFs, fnFs.filtN, fnRs, fnRs.filtN, maxN = 0, multithread = TRUE) 
# CHANGE multithread to FALSE on Windows (here and elsewhere in the program)

```

















